packages/language/src/preprocessor/instruction-interpreter.ts

```diff
@@ -319,38 +324,235 @@ function runInstruction(
 function runDoInstruction(
   instruction: inst.DoInstruction,
   context: InterpreterContext,
+  node?: inst.InstructionNode,
 ): inst.InstructionNode | undefined {
   let condition = true;
   if (instruction.doType2) {
-    const type2 = instruction.doType2;
-    if (type2.until) {
-      const untilConditionValue = evaluateExpression(type2.until, context);
-      if (!isScalarValue(untilConditionValue)) {
-        // Condition cannot be evaluated, don't run the instruction
-        condition = false;
-      } else {
-        // If UNTIL is specified, we don't run the instruction if it evaluates to true
-        condition &&= !valueToBool(untilConditionValue);
-      }
+    condition = runDoType2Instruction(instruction.doType2, context);
+  } else if (instruction.doType3) {
+    if (!node) {
+      throw new Error(
+        "DoType3 instruction requires a node for iteration tracking",
+      );
+    }
+    condition = runDoType3Instruction(instruction.doType3, context, node);
+  }
+  if (condition) {
+    return instruction.content;
+  }
+  return undefined;
+}
+
+function runDoType2Instruction(
+  doType2: inst.DoType2Instruction,
+  context: InterpreterContext,
+): boolean {
+  let condition = true;
+  if (doType2.until) {
+    const untilConditionValue = evaluateExpression(doType2.until, context);
+    if (!isScalarValue(untilConditionValue)) {
+      // Condition cannot be evaluated, don't run the instruction
+      condition = false;
+    } else {
+      // If UNTIL is specified, we don't run the instruction if it evaluates to true
+      condition &&= !valueToBool(untilConditionValue);
+    }
+  }
+  if (doType2.while) {
+    const whileConditionValue = evaluateExpression(doType2.while, context);
+    if (!isScalarValue(whileConditionValue)) {
+      // Condition cannot be evaluated, don't run the instruction
+      condition = false;
+    } else {
+      // If WHILE is specified, we only run the instruction if it evaluates to true
+      condition &&= valueToBool(whileConditionValue);
+    }
+  }
+  return condition;
+}
+
+function runDoType3Instruction(
+  doType3: inst.DoType3Instruction,
+  context: InterpreterContext,
+  node: inst.InstructionNode,
+): boolean {
+  const { specificationItems, variable } = doType3;
+
+  if (specificationItems.length === 0) {
+    throw new Error("DoType3 requires at least one specification item!");
+  }
+
+  const iterationCount = context.counter.get(node);
+  if (iterationCount === undefined || iterationCount < 1) {
+    throw new Error("Interpreter can't determine AST node iteration count");
+  }
+  const isFirstIteration = iterationCount === 1;
+
+  // Get current spec index, starting at 0 for first iteration
+  let currentSpecIndex = context.doType3SpecIndex.get(node) ?? 0;
+  const needsReinit = context.doType3NeedsReinit.get(node) ?? false;
```

Comment:
Suggestion: I don't think we need the iteration count at all here, if we set the default for `needsReinit` differently.
```suggestion
  // Get current spec index, starting at 0 for first iteration
  let currentSpecIndex = context.doType3SpecIndex.get(node) ?? 0;
  const needsReinit = context.doType3NeedsReinit.get(node) ?? true;
```